metals/src/main/scala/scala/meta/internal/parsing/JavaTokens.scala

```diff
@@ -0,0 +1,159 @@
+package scala.meta.internal.parsing
+
+import scala.meta.inputs.Input
+import scala.util.Try
+import org.eclipse.jdt.core.ToolFactory
+import org.eclipse.jdt.core.compiler.ITerminalSymbols
+import scala.meta.tokens.Token
+import scala.meta.dialects
+import org.eclipse.jdt.core.compiler.IScanner
+import scala.meta.tokens.Tokens
+import scala.collection.mutable.ArrayBuffer
+import scala.meta.Dialect
+
+object JavaTokens {
+
+  /**
+   * Best effort translation of java tokens to the scalameta tokens, should work for most purposes.
+   *
+   * @param source to tokenize
+   * @return tokenized source
+   */
+  def tokenize(source: Input): Try[Tokens] = Try {
+
+    val scanner = ToolFactory.createScanner(
+      /*tokenizeComments = */ true, /*tokenizeWhiteSpace = */ true,
+      /*assertMode = */ false, /*recordLineSeparator = */ true
+    )
+    scanner.setSource(source.chars)
+
+    val buffer = new ArrayBuffer[Token]()
+
+    def loop(token: Int): Unit = {
+      if (token != ITerminalSymbols.TokenNameEOF) {
+        mapToken(token, scanner, source, buffer)
+        loop(scanner.getNextToken())
+      }
+    }
+
+    loop(scanner.getNextToken())
+    val array = buffer.toArray
+    Tokens(array, 0, array.size)
+  }
+
+  private def mapToken(
+      i: Int,
+      scanner: IScanner,
+      source: Input,
+      buffer: ArrayBuffer[Token]
+  ): Unit = {
```

Comment:
`getToken` should return token which will be added to the buffer in `loop`. This will simplify implementation of `token`, `keyword`, `constant` since they will be doing just one thing -- creating token.
```suggestion
  private def getToken(
      i: Int,
      scanner: IScanner,
      source: Input
  ): Token = {
```